ksqldb-engine/src/main/java/io/confluent/ksql/serde/SerdeFeaturesFactory.java

```diff
@@ -150,6 +161,20 @@ private static KeyFormat sanitizeKeyFormatForMultipleColumns(
     return KeyFormat.nonWindowed(FormatInfo.of(JsonFormat.NAME), SerdeFeatures.of());
   }
 
+  /**
+   * Check whether given sql type set needs a compound key support. It returns true
+   * when one of the conditions is met:
+   * 1. given sql types have more than one entry
+   * 2. it is a singleton non-primitive sql type
+   *
+   * @param sqlTypes the sql types to be tested
+   * @return true if given sql types do not need a compound key type support.
+   */
+  private static boolean needCompoundKeySupport(List<SqlType> sqlTypes) {
+    return sqlTypes.size() > 1 || (sqlTypes.size() == 1 &&
```

Comment:
I don't think it makes sense to combine these two conditions. As per my comment above, there are two different properties that are separate: whether a format supports multiple columns (e.g., two string columns) and whether a format supports complex-types (e.g., struct or array). The two properties are different as the delimited format supports the former but not the latter. We need two separate checks in the methods in this file: (1) the original check for multiple columns (which is correct and should not need to be changed), and (2) a new check for whether specific column types are supported (what's currently missing, and why the struct type is causing a bug).